lib/badge_queries.rb

```diff
@@ -272,4 +272,31 @@ def self.consecutive_visits(days)
     SQL
   end
 
+  def self.anniversaries(start_date, end_date)
+    start_date = start_date.iso8601(6)
+    end_date = end_date.iso8601(6)
+
+    <<~SQL
+      SELECT u.id
+        FROM users AS u
+        JOIN posts AS p ON p.user_id = u.id
+        JOIN topics AS t ON p.topic_id = t.id
+       WHERE u.id > 0
+         AND u.active
+         AND NOT u.staged
+         AND (u.silenced_till IS NULL OR u.silenced_till < '#{start_date}')
+         AND (u.suspended_till IS NULL OR u.suspended_till < '#{start_date}')
+         AND u.created_at <= '#{start_date}'
+         AND NOT p.hidden
+         AND p.deleted_at IS NULL
+         AND p.created_at BETWEEN '#{start_date}' AND '#{end_date}'
+         AND t.visible
+         AND t.archetype <> 'private_message'
+         AND t.deleted_at IS NULL
+         AND NOT EXISTS (SELECT 1 FROM user_badges AS ub WHERE ub.user_id = u.id AND ub.badge_id = #{Badge::Anniversary} AND ub.granted_at BETWEEN '#{start_date}' AND '#{end_date}')
```

Comment:
Is there any difference between this and joining on the user_badges table?

Reply:
There is no behaviour between doing 

```sql
LEFT JOIN user_badges AS ub ON ub.user_id = u.id AND ...
[...]
HAVING COUNT(ub.*) = 0
```

and

```sql
NOT EXISTS (SELECT 1 FROM user_badges AS ub WHERE ub.user_id = u.id AND ...)
```

The only difference is performance-wise. The `NOT EXISTS` is 10-30% faster on very large databases (aka. posts and users in X millions). I checked on 3 of the largest datasets I could find.